nexus/src/authz/api_resources.rs

```diff
@@ -0,0 +1,303 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Authz types for resources in the API hierarchy
+
+// Most of the types here are used in the Polar configuration, which means they
+// must impl [`oso::PolarClass`].  There is a derive(PolarClass), but it's
+// pretty limited: It also doesn't let you define methods.  We
+// may want to define our own macro(s) to avoid having to impl this by hand
+// everywhere.
+//
+// Many of these are newtypes that wrap existing types, either simple types from
+// Nexus (like `authn::Context`) or database model types.  The main reason is
+// that [`Oso::is_allowed()`] consumes owned values, not references.  Consumers
+// that want to do an authz check almost always want to keep using the thing
+// they had.  We could instead require that the database model types all impl
+// `Clone`, impl `PolarClass` on those types, and then just copy them when we
+// pass them to `is_allowed()`.  Using newtypes is a way to capture just the
+// parts we need for authorization.
+
+use super::actor::AuthenticatedActor;
+use super::roles::AuthzApiResource;
+use super::roles::AuthzResource;
+use crate::db::fixed_data::FLEET_ID;
+use omicron_common::api::external::ResourceType;
+use uuid::Uuid;
+
+/// Represents the Oxide fleet for authz purposes
+///
+/// Fleet-level resources are essentially global.  See RFD 24 for more on
+/// Fleets.
+///
+/// This object is used for authorization checks on a Fleet by passing it as the
+/// `resource` argument to [`crate::context::OpContext::authorize()`].  You
+/// typically don't construct a `Fleet` yourself -- use the global [`FLEET`].
+///
+/// You can perform authorization checks on children of the Fleet (e.g.,
+/// Organizations) using the methods below that return authz objects
+/// representing those children.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Fleet;
+/// Singleton representing the [`Fleet`] itself for authz purposes
+pub const FLEET: Fleet = Fleet;
+
+impl Fleet {
+    /// Returns an authz resource representing a child Organization
+    pub fn organization(&self, organization_id: Uuid) -> Organization {
+        Organization { organization_id }
+    }
+
+    /// Returns an authz resource representing some other kind of child (e.g.,
+    /// a built-in user, built-in role, etc. -- but _not_ an Organization)
+    ///
+    /// Aside from Organizations (which you create with [`Fleet::organization()`
```

Comment:
nit: missing a closing `]` here.

Reply:
Fixed in 00ac778eadf087dc0be1d8d652dbeee55521189a.